ublox/src/parser/ubx_rtcm.rs

```diff
@@ -0,0 +1,162 @@
+//! Adaptative UBX + RTCM parser
+
+use crate::{
+    error::ParserError,
+    parser::{ubx::extract_packet_ubx, DualBuffer, UnderlyingBuffer},
+    ubx_packets::{PacketRef, MAX_PAYLOAD_LEN, RTCM_SYNC_CHAR, SYNC_CHAR_1, SYNC_CHAR_2},
+};
+
+#[cfg(feature = "rtcm")]
+use rtcm_rs::{next_msg_frame as find_next_rtcm_frame, MessageFrame as RtcmMessageFrame};
+
+#[derive(Debug, PartialEq, Eq)]
+enum NextSync {
+    Ubx(usize),
+    Rtcm(usize),
+    None,
+}
+
+/// [AnyPacketRef] allows identifying UBX [PacketRef]s and RTCM packets on the fly.
+pub enum AnyPacketRef<'a> {
+    Ubx(PacketRef<'a>),
+
+    #[cfg(feature = "rtcm")]
+    Rtcm(RtcmMessageFrame<'a>),
+
+    #[cfg(not(feature = "rtcm"))]
+    /// Reference to underlying RTCM bytes, "as is".
+    Rtcm(RtcmPacketRef<'a>),
+}
+
+/// Iterator over data stored in `Parser` buffer. Both UBX and RTCM
+/// packets will be identified on the fly.
+pub struct UbxRtcmParserIter<'a, T: UnderlyingBuffer> {
+    pub(crate) buf: DualBuffer<'a, T>,
+}
+
+/// Reference to RTCM content, "as is".
+/// When compiled with `rtcm` feature, we can forward this content to
+/// the RTCM parser.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct RtcmPacketRef<'a> {
+    pub data: &'a [u8],
+}
+
+fn extract_packet_rtcm<'a, 'b, T: UnderlyingBuffer>(
+    buf: &'b mut DualBuffer<'a, T>,
+    pack_len: usize,
+) -> Option<Result<AnyPacketRef<'b>, ParserError>> {
+    if !buf.can_drain_and_take(0, pack_len + 3) {
+        if buf.potential_lost_bytes() > 0 {
+            // We ran out of space, drop this packet and move on
+            // TODO: shouldn't we drain pack_len + 3?
+            buf.drain(2);
+            return Some(Err(ParserError::OutOfMemory {
+                required_size: pack_len + 2,
+            }));
+        }
+        return None;
+    }
+
+    let maybe_data = buf.take(pack_len + 3);
+
+    let (consumed, msg_frame) = match maybe_data {
+        Ok(data) => find_next_rtcm_frame(data),
+        Err(e) => {
+            return Some(Err(e));
+        },
+    };
+
+    if consumed > 0 {
+        // remove bytes that have been consumed by RTCM identification attempt
```

Comment:
If the RTCM decoder did consume something, ideally we should advance within our buffer